compiler/noirc_evaluator/src/ssa/opt/die.rs

```diff
@@ -1179,4 +1206,60 @@ mod test {
         let ssa = ssa.dead_instruction_elimination();
         assert_normalized_ssa_equals(ssa, src);
     }
+
+    #[test]
+    fn remove_all_instructions_in_unreachable_block_but_final_trap() {
+        let src = "
+        acir(inline) predicate_pure fn main f0 {
+          b0():
+            v0 = allocate -> &mut u1                            
+            v2 = make_array [u1 0, v0] : [(u1, &mut u1); 1]     
+            v4 = array_get v2, index u32 2 -> u1                
```

Comment:
Shouldn't the current DIE pass (without the changes in this PR) already remove these as unused?

Reply:
I am going to branch off of #9232 as this is mostly relevant for that case, and I'm not sure if this is the best solution anymore.